crates/core/tedge_agent/src/file_transfer_server/entity_store.rs

```diff
@@ -0,0 +1,409 @@
+use super::http_rest::AgentState;
+use crate::entity_manager::server::EntityStoreRequest;
+use crate::entity_manager::server::EntityStoreResponse;
+use axum::extract::Path;
+use axum::extract::State;
+use axum::response::IntoResponse;
+use axum::response::Response;
+use axum::routing::post;
+use axum::Json;
+use axum::Router;
+use hyper::StatusCode;
+use serde::Deserialize;
+use serde::Serialize;
+use std::str::FromStr;
+use tedge_actors::Sender;
+use tedge_api::entity_store;
+use tedge_api::entity_store::EntityMetadata;
+use tedge_api::entity_store::EntityRegistrationMessage;
+use tedge_api::mqtt_topics::Channel;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::TopicIdError;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::QoS;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub(crate) struct EntityDef {
+    topic_id: String,
+}
+
+#[derive(thiserror::Error, Debug)]
+enum Error {
+    #[error(transparent)]
+    InvalidEntityTopicId(#[from] TopicIdError),
+
+    #[allow(clippy::enum_variant_names)]
+    #[error(transparent)]
+    EntityStoreError(#[from] entity_store::Error),
+
+    #[error("Entity not found with topic id: {0}")]
+    EntityNotFound(EntityTopicId),
+
+    #[allow(clippy::enum_variant_names)]
+    #[error("Failed to publish entity registration message via MQTT")]
+    ChannelError(#[from] tedge_actors::ChannelError),
+
+    #[error("Received unexpected response from entity store")]
+    InvalidEntityStoreResponse,
+}
+
+impl IntoResponse for Error {
+    fn into_response(self) -> Response {
+        let status_code = match &self {
+            Error::InvalidEntityTopicId(_) => StatusCode::BAD_REQUEST,
+            Error::EntityStoreError(_) => StatusCode::BAD_REQUEST,
+            Error::EntityNotFound(_) => StatusCode::NOT_FOUND,
+            Error::ChannelError(_) => StatusCode::INTERNAL_SERVER_ERROR,
+            Error::InvalidEntityStoreResponse => StatusCode::INTERNAL_SERVER_ERROR,
+        };
+        let error_message = self.to_string();
+
+        (status_code, error_message).into_response()
+    }
+}
+
+pub(crate) fn entity_store_router(state: AgentState) -> Router {
+    Router::new()
+        .route(
+            "/v1/entities/*path",
+            post(register_entity)
+                .get(get_entity)
+                .delete(deregister_entity),
+        )
+        .with_state(state)
+}
+
+async fn register_entity(
+    State(state): State<AgentState>,
+    Json(entity): Json<EntityRegistrationMessage>,
+) -> Result<StatusCode, Error> {
+    let response = state
+        .entity_store_handle
+        .clone()
+        .await_response(EntityStoreRequest::Create(entity.clone()))
+        .await?;
+    let EntityStoreResponse::Create(result) = response else {
+        return Err(Error::InvalidEntityStoreResponse);
+    };
+
+    if !result?.0.is_empty() {
+        let message = entity.to_mqtt_message(&state.mqtt_schema);
+        state.mqtt_publisher.clone().send(message).await?;
+    }
```

Comment:
I would delegate to `EntityStoreServer` the responsibility to publish over MQTT the updated entity metadata.

-  `EntityStoreServer`  is already MQTT dependent and publishing entity metadata.
- This HTTP handler would be free of any MQTT dependencies.

Reply:
Resolved by a85933b